jobsdb/migration.go

```diff
@@ -0,0 +1,472 @@
+			}
+			err = jd.journalMarkDoneInTx(tx, opID)
+			if err != nil {
+				return err
+			}
+			jd.logger.Infof("[[ migrateDSLoop ]]: Total migrated %d jobs", totalJobsMigrated)
+		}
+
+		opPayload, err := json.Marshal(&journalOpPayloadT{From: migrateFrom})
+		if err != nil {
+			return err
+		}
+		opID, err := jd.JournalMarkStartInTx(tx, postMigrateDSOperation, opPayload)
+		if err != nil {
+			return err
+		}
+		// acquire an async lock, as this needs to be released after the transaction commits
+		l, lockChan, err = jd.dsListLock.AsyncLockWithCtx(ctx)
+		if err != nil {
+			return err
+		}
+		err = jd.postMigrateHandleDS(tx, migrateFrom)
+		if err != nil {
+			return err
+		}
+		return jd.journalMarkDoneInTx(tx, opID)
+	})
+	if l != nil {
+		if err == nil {
+			jd.refreshDSRangeList(l)
+		}
+		lockChan <- l
+	}
+	return err
+}
+
+// based on an estimate of the rows in DSs, gives a list of DSs for us to cleanup status tables
+func (jd *HandleT) getCleanUpCandidates(ctx context.Context, dsList []dataSetT) ([]dataSetT, error) {
+	// get analyzer estimates for the number of rows(jobs, statuses) in each DS
+	var rows *sql.Rows
+	rows, err := jd.dbHandle.QueryContext(
+		ctx,
+		`SELECT reltuples AS estimate, relname 
+		FROM pg_class 
+		where relname = ANY(
+			SELECT tablename 
+				FROM pg_catalog.pg_tables 
+				WHERE schemaname NOT IN ('pg_catalog','information_schema')
+				AND tablename like $1
+		)`,
+		jd.tablePrefix+"_job%",
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = rows.Close() }()
+
+	estimates := map[string]int64{}
+	for rows.Next() {
+		var (
+			estimate  int64
+			tableName string
+		)
+		err = rows.Scan(&estimate, &tableName)
+		if err != nil {
+			return nil, err
+		}
+		estimates[tableName] = estimate
+	}
+
+	datasets := lo.Filter(dsList,
+		func(ds dataSetT, idx int) bool {
+			statuses := estimates[ds.JobStatusTable]
+			jobs := estimates[ds.JobTable]
+			if jobs == 0 { // using max ds size if we have no stats for the number of jobs
+				jobs = int64(*jd.MaxDSSize)
+			}
+			return float64(statuses)/float64(jobs) > jobStatusMigrateThres
+		})
+
+	return lo.Slice(datasets, 0, maxMigrateDSProbe), nil
+}
+
+// based on an estimate cleans up the status tables
+func (jd *HandleT) cleanupStatusTables(ctx context.Context, dsList []dataSetT) error {
+	toCompact, err := jd.getCleanUpCandidates(ctx, dsList)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	defer stats.Default.NewTaggedStat(
+		"jobsdb_compact_status_tables",
+		stats.TimerType,
+		stats.Tags{"customVal": jd.tablePrefix},
+	).Since(start)
+
+	return jd.WithTx(func(tx *Tx) error {
+		for _, statusTable := range toCompact {
+			if err := jd.cleanStatusTable(
+				ctx,
+				tx,
+				statusTable.JobStatusTable,
+			); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// cleanStatusTable deletes all rows except for the latest status for each job
+func (*HandleT) cleanStatusTable(ctx context.Context, tx *Tx, table string) error {
```

Comment:
let's also run an `ANALYZE` on the status table after the delete, since we will be deleting a lot of record.